pallets/stateful-storage/src/types.rs

```diff
@@ -19,41 +19,108 @@ pub struct ItemHeader {
 	payload_len: u16,
 }
 
+/// This header is used to specify how much data is in a page and is inserted into the buffer
+/// before every page in a Paginated model.
+#[derive(Clone, Encode, Decode, PartialEq, MaxEncodedLen, Debug)]
+pub struct PageHeader {
+	/// The length of this item, not including the size of this header.
+	payload_len: u16,
+}
+
 #[derive(Debug, PartialEq)]
-pub enum ItemPageError {
+pub enum PageError {
 	ErrorParsing(&'static str),
 	InvalidAction(&'static str),
 	ArithmeticOverflow,
+	PageSizeOverflow,
 }
 
-/// A page of items
+// NOTE: Would prefer to simply have an enum:
+// ```no_compile
+// enum PageType { Itemized, Paginated }
+// ```
+// However, currently can't use enum variants as const generic parameters.
+pub trait PageType {}
+
+#[derive(Debug)]
+struct Itemized;
+
+#[derive(Debug)]
+struct Paginated;
+
+impl PageType for Itemized {}
+impl PageType for Paginated {}
+
+/// A page of data
 #[derive(Encode, Decode, TypeInfo, MaxEncodedLen, PartialEq, Debug)]
 #[scale_info(skip_type_params(PageDataSize))]
 #[codec(mel_bound(PageDataSize: MaxEncodedLen))]
-pub struct ItemPage<PageDataSize: Get<u32>> {
+pub struct Page<P: PageType, PageDataSize: Get<u32>> {
```

Comment:
Pages can not be under the same Type for a simple reason. `PageDataSize` for itemized pages are different from Paginated ones. I think we will need to have two different Types and we can implement the same Trait on both if we like to treat them as general pages.

Reply:
But since `Page` is a generic, we can have 2 concrete types:
```
type PaginatedPage = Page<Paginated, MaxPaginatedPageSizeBytes>;
type ItemizedPage = Page<Itemized, MaxItemizedPageSizeBytes>;
```

Reply:
I like the different types from common generic, if you can make it work go for it.

I don't think they way you are using the paginated and itemized as first generic to differentiate between them is idiomatic Rust but I have to play with it to see how else we might be able to do it.

Reply:
There are other problem with a generic inner type. The macros we use on each of them might be different like 
`#[codec(mel_bound(T::MaxItemizedPageSizeBytes: MaxEncodedLen))]`

I'm not sure how we would be able to apply this macros based on type

Reply:
Okay, reworked a bit. Once I eliminated `PageHeader` as unnecessary, there was also no longer a need for `ParsedPaginatedPage`. This simplifies things quite a bit.